core/modules/replica/WorkerServerConnection.h

```diff
@@ -0,0 +1,225 @@
+     *        how many bytes exactly are we going to read.
+     * 
+     * The chain ends when a client disconnects or when an error condition
+     * is met.
+     */
+    void beginProtocol();
+
+private:
+
+    /**
+     * The constructor of the class
+     *
+     * @see WorkerServerConnection::create()
+     */
+    WorkerServerConnection(ServiceProvider::Ptr const& serviceProvider,
+                           WorkerProcessor& processor,
+                           boost::asio::io_service& io_service);
+
+    /**
+     * Begin reading (asynchronosly) the frame header of a new request
+     *
+     * The frame header is presently a 32-bit unsigned integer
+     * representing the length of the subsequent message.
+     */
+    void receive();
+
+    /**
+     * The calback on finishing (either successfully or not) of aynchronious reads.
+     *
+     * @param ec                - error condition to be checked for
+     * @param bytes_transferred - the number of bytes receoved (if successful)
+     */
+    void received(boost::system::error_code const& ec,
+                  size_t bytes_transferred);
+
+    /**
+     * Process replication requests (REPLICATE, DELETE, FIND, FIND-ALL)
+     *
+     * @param hdr - request header to be inspected
+     */
+    void processReplicaRequest(proto::ReplicationRequestHeader& hdr);
+
+    /**
+     * Process requests about replication requests (STOP, STATUS)
+     *
+     * @param hdr - request header to be inspected
+     */
+    void processManagementRequest(proto::ReplicationRequestHeader& hdr);
+
+    /**
+     * Process requests affecting the service
+     *
+     * @param hdr - request header to be inspected
+     */
+    void processServiceRequest(proto::ReplicationRequestHeader& hdr);
+
+    /**
+     * Serialize an identifier of a request into response header
+     * followed by the protobuf response body protobuf object and
+     * send it all back to a client.
+     *
+     * @param id   - a unique identifier of a request to which th ereply is sent
+     * @param body - a body of the response
+     */
+    template <class T>
+    void reply(std::string const& id,
+               T&& body) {
+
+        _bufferPtr->resize();
+
+        proto::ReplicationResponseHeader hdr;
+        hdr.set_id(id);
+
+        _bufferPtr->serialize(hdr);
+        _bufferPtr->serialize(body);
+
+        send();
+    }
+
+    /**
+     * Begin sending (asynchronosly) a result back to a client
+     */
+    void send();
+
+    /**
+     * The calback on finishing (either successfully or not) of aynchronious writes.
+     *
+     * @param ec                - error condition to be checked for
+     * @param bytes_transferred - the number of bytes sent (if successful)
+     */
+    void sent(boost::system::error_code const& ec,
+              size_t bytes_transferred);
+
+private:
+
+    ServiceProvider::Ptr _serviceProvider;
+    WorkerProcessor& _processor;
```

Comment:
you're sure the WorkerProcessor isn't going to get deleted while this object still exists? 
you might consider using a shared_ptr or weak_ptr instead of keeping a simple reference.

Reply:
In the current implementation it shouldn't because it's a reference to a member variable of class WorkerServer which itself gets instantiated in the application qserv-replica-worker.cc. However, I can see certain risks with that in case if someone makes some modifications to the code. I will investigate a possibility of turning that reference into a smart pointer.
Thanks again!